src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -401,24 +401,30 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
   }
 
   void push(const E& elem) { append(elem); }
-
-  E at_grow(int i, const E& fill = E()) {
+private:
+  static void default_fill(E* ptr) {
+    new ((void*)ptr) E();
+  }
+public:
+  template<typename... Args >
+  E at_grow(int i, Args... args) {
```

Comment:
Should this be `Args&... args`? Also this method is returning a value instead of a reference, is it intentional?

Reply:
>Should this be Args&... args?

Yes.

>Also this method is returning a value instead of a reference, is it intentional?

That is difficult to say. The caller can decided whether to capture the value by-reference or by-value either way.

Reply:
> > Should this be Args&... args?
> 
> Yes.
> 

Actually, no: We can't take by reference because most of the time we're calling `at_grow` and `at_put_grow` with rvalues and not lvalues, so we need move semantics for this to work. We have to keep the arguments as copying.

Reply:
I see, in that case it must be `const Args& args`. https://godbolt.org/z/aov8v38s9

Reply:
Fixed, thanks for the help with this.

Reply:
For perfect forwarding, `Args&&...` should be used together with `std::forward` when doing the downstream call to the constructor. That will make it so rvalue arguments are forwarded as rvalues as well. The current code turns all args into lvalues.

Do we need to support NONCOPYABLE argument types as well?

Reply:
Hi Jorn, the style guide currently doesn't allow for using perfect forwarding or move semantics in the code, so we cannot do this at the moment.

There's a PR for that: https://github.com/openjdk/jdk/pull/15386

Reply:
Ah, ok I see. Thanks for the link.

Reply:
This is a case where using rvalues and perfect forwarding is incorrect. The
arguments are potentially used in multiple E constructor calls. Using the same
rvalue in multiple places is very bad.

Reply:
Ah, that's right. We call it in a loop, so `args...` might be used multiple times.